File: examples/spaced/ts/commands/competitor.ts
/**
 * commands/competitor.ts — competitor compare <name> --metric
 */

import { Command } from 'commander';
import { findCompetitor } from '../data';

export function competitorCommand(): Command {
  const competitor = new Command('competitor')
    .description('Compare SpaceX against its competitors');

  competitor
    .command('compare <name>')
    .description('Compare a competitor to SpaceX')
    .option('--metric <value>', 'Comma-separated metrics to show (default: all)', '')
    .action(function (name: string) {
      const opts = this.opts();
      const c = findCompetitor(name);

      if (!c) {
        process.stderr.write(`  error: unknown competitor: "${name}"\n`);
        const known = ['Boeing', 'Blue Origin', 'Virgin Galactic', 'ULA', 'Roscosmos'];
        process.stderr.write(`  Known competitors: ${known.join(', ')}\n`);
        process.exit(1);
      }

      const filterMetrics = opts['metric']
        ? (opts['metric'] as string).split(',').map((m: string) => m.trim().toLowerCase()).filter(Boolean)
        : [];

      console.log('');
      console.log(`  COMPETITOR: ${c.name}`);
      console.log(`  ${'─'.repeat(60)}`);
      console.log('');
      console.log(`  Founded:              ${c.founded}`);
      console.log(`  Active vehicles:      ${c.rockets.join(', ')}`);
      console.log(`  Crewed flights:       ${c.crewed_flights}`);
      console.log(`  Success rate:         ${c.launch_success_rate}`);
      console.log('');
      console.log(`  Notable achievement:  ${c.notable_achievement}`);
      console.log(`  Notable failure:      ${c.notable_failure}`);
      console.log('');
      console.log(`  Elon's opinion:       ${c.elon_opinion}`);
      console.log('');

      const metricKeys = filterMetrics.length > 0
        ? Object.keys(c.metrics).filter(k =>
          filterMetrics.some(f => k.toLowerCase().includes(f)),
        )
        : Object.keys(c.metrics);

      if (metricKeys.length > 0) {
        console.log('  Key metrics:');
        for (const key of metricKeys) {
          const label = key.replace(/_/g, ' ').padEnd(28);
          console.log(`    ${label} ${c.metrics[key]}`);
        }
        console.log('');
      }

      console.log('  vs SpaceX:');
      console.log('    SpaceX Falcon 9:    $2,600/kg to LEO (reuse amortized)');
      console.log('    SpaceX Starship:    ~$100/kg to LEO (target; unverified)');
      console.log('    SpaceX crewed:      14 Dragon crew flights; all successful');
      console.log('');
      console.log('  Disclaimer: This comparison is satirical. SpaceX may also have');
      console.log('  failures we have chosen not to highlight out of narrative convenience.');
      console.log('');
    });

  return competitor;
}

File: examples/spaced/py/commands/competitor.py
"""commands/competitor.py — competitor compare <name>."""
from __future__ import annotations

import sys
from typing import Optional

import typer

sys.path.insert(0, __import__("os").path.join(__import__("os").path.dirname(__file__), "../../../../sdk/py"))
from data import COMPETITORS, find_competitor  # noqa: E402

app = typer.Typer(help="Compare SpaceX against its competitors", add_completion=False)


@app.command("compare")
def competitor_compare(
    name: str = typer.Argument(..., help="Competitor name (Boeing, Blue Origin, ULA, ...)"),
    metric: Optional[str] = typer.Option(
        None, "--metric", help="Comma-separated metrics to show (e.g. reliability,cost_per_kg)"
    ),
) -> None:
    """Compare a competitor against SpaceX. Spoiler: SpaceX wins."""
    c = find_competitor(name)
    if c is None:
        available = ", ".join(comp.name for comp in COMPETITORS)
        typer.echo(f"  ✗ Competitor not found: {name}", err=True)
        typer.echo(f"  Available: {available}", err=True)
        raise typer.Exit(1)

    metric_filter = [m.strip().lower() for m in metric.split(",")] if metric else []

    rockets_str = ", ".join(c.rockets)
    crewed_str = "yes" if c.crewed else "no"

    typer.echo(f"""
  COMPETITOR ANALYSIS: {c.name}
  {"─" * 56}
  Founded:    {c.founded}
  Status:     {c.status}
  Tagline:    {c.tagline}
  Rockets:    {rockets_str}
  Crewed:     {crewed_str}

  Notable failure:
    {c.notable_failure}

  Metrics (vs SpaceX: dominant):
""")

    for key, val in c.metrics.items():
        if metric_filter and key.lower() not in metric_filter:
            continue
        typer.echo(f"    {key:<28}{val}")

    typer.echo(f"""
  SpaceX comparison:
    Launches/year:  50+ (SpaceX) vs. {_cadence_guess(c.name)}
    Cost/kg (LEO):  ~$2,700 (F9) — competitors: more
    Reusability:    yes — competitors: mostly no

  Verdict: {_verdict(c.name)}
""")


def _cadence_guess(name: str) -> str:
    return {
        "Boeing": "aspirational (Starliner still on pad, spiritually)",
        "Blue Origin": "~6/year (improving)",
        "Virgin Galactic": "0 (suspended)",
        "ULA": "~6/year",
        "Roscosmos": "~15/year (declining)",
    }.get(name, "fewer")


def _verdict(name: str) -> str:
    return {
        "Boeing": "Historically great. Currently: cautionary tale. Root cause: culture.",
        "Blue Origin": "Better than headline. Worse than SpaceX. Bezos still funding.",
        "Virgin Galactic": "Pioneered space tourism concept. Currently: not operating it.",
        "ULA": "Reliable, expensive, government-dependent. Vulcan: the sequel.",
        "Roscosmos": "Gagarin would not recognize it. Sanctions didn't help.",
    }.get(name, "Competing. That's something.")
